gated_filter.go

```diff
@@ -0,0 +1,364 @@
+package eventlogger
+
+import (
+	"container/list"
+	"context"
+	"fmt"
+	"sync"
+	"time"
+)
+
+// Gateable defines an interface for Event payloads which are "gateable" by
+// the GatedFilter
+type Gateable interface {
+	// GetID returns an ID which allows the GatedFilter to determine that the
+	// payload is part of a group of Gateable payloads.
+	GetID() string
+
+	// FlushEvent returns true when the Gateable event payload includes a Flush
+	// indicator.
+	FlushEvent() bool
+
+	// ComposeFrom creates one payload which is a composition of a list events.
+	// When ComposeFrom(...) is called by a GatedFilter the receiver will
+	// always be nil. The payload returned must not have a Gateable payload.
+	ComposeFrom(events []*Event) (t EventType, payload interface{}, err error)
+}
+
+// gatedEvent is a list of Events with the same Gateable.GetID().  These events
+// have an exp (expiration) time. A gatedEvent has a list.Element, which allows
+// it to be part of a linked list of gatedEvents.
+type gatedEvent struct {
+	// id of the event and all the "events"
+	id string
+	// events are an ordered list that all have the same ID
+	events []*Event
+	// exp of the gatedEvent
+	exp time.Time
+	// element of a linked list of gatedEvents
+	element *list.Element
+}
+
+// DefaultGatedEventTimeout defines a default expiry for events processed by a
+// GatedFilter
+const DefaultGatedEventTimeout = time.Second * 10
+
+// GatedFilter provides the ability to buffer events identified by a
+// Gateable.GetID() until an event is processed that returns true for
+// Gateable.FlushEvent().
+//
+// When a Gateable Event returns true for FlushEvent(), the filter will call
+// Gateable.ComposedOf(...) with the list of gated events with the coresponding
+// Gateable.GetID() up to that point in time and return the resulting composed
+// event.
+//
+// If GatedFilter.Broker is nil, expired gated events will simply be deleted.
```

Comment:
`If GatedFilter.Broker is nil, expired gated events will simply be deleted.`
If GatedFilter.Broker is nil and the event is not expired, what happens? I'm assuming if it does not have a broker then nothing?

Reply:
These two comments are about dependencies on the GatedFilter.Broker when there are expiring events.   There is no dependency on the GatedFilter.Broker for handling non-expired and non-flushed events.... they just stay in the same state until they expire or their flushed.... then there's the documented dependency on the Broker and how those expired events are handled and there's no dependency on the broker to flush an event.

Reply:
I'll try to add something to be more clear... but I'm not convinced it will help to try and define all the "non-dependent" states... 🤷

Reply:
I think the comments I've added will be helpful... ty!